core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpParser.java

```diff
@@ -0,0 +1,635 @@
+    public static final byte ENTITY_TYPE_NULL = 0;
+    public static final byte ENTITY_TYPE_TAG = 1;
+    public static final byte ENTITY_TYPE_FLOAT = 2;
+    public static final byte ENTITY_TYPE_INTEGER = 3;
+    public static final byte ENTITY_TYPE_STRING = 4;
+    public static final byte ENTITY_TYPE_SYMBOL = 5;
+    public static final byte ENTITY_TYPE_BOOLEAN = 6;
+    public static final byte ENTITY_TYPE_LONG256 = 7;
+    public static final byte ENTITY_TYPE_CACHED_TAG = 8;
+    public static final byte ENTITY_TYPE_GEOBYTE = 9;
+    public static final byte ENTITY_TYPE_GEOSHORT = 10;
+    public static final byte ENTITY_TYPE_GEOINT = 11;
+    public static final byte ENTITY_TYPE_GEOLONG = 12;
+    public static final byte ENTITY_TYPE_TIMESTAMP = 13;
+    public static final int N_ENTITY_TYPES = ENTITY_TYPE_TIMESTAMP + 1;
+    static final byte ENTITY_TYPE_NONE = (byte) 0xff; // visible for testing
+    private final DirectByteCharSequence measurementName = new DirectByteCharSequence();
+    private final DirectByteCharSequence charSeq = new DirectByteCharSequence();
+    private final ObjList<ProtoEntity> entityCache = new ObjList<>();
+    private final EntityHandler entityEndOfLineHandler = this::expectEndOfLine;
+    private long bufAt;
+    private long entityLo;
+    private boolean tagsComplete;
+    private int nEscapedChars;
+    private boolean isQuotedFieldValue;
+    private int nEntities;
+    private ProtoEntity currentEntity;
+    private ErrorCode errorCode;
+    private EntityHandler entityHandler;
+    private final EntityHandler entityTableHandler = this::expectTableName;
+    private long timestamp;
+    private final EntityHandler entityTimestampHandler = this::expectTimestamp;
+    private final EntityHandler entityValueHandler = this::expectEntityValue;
+    private final EntityHandler entityNameHandler = this::expectEntityName;
+    private int nQuoteCharacters;
+    private boolean scape;
+    private boolean nextValueCanBeOpenQuote;
+    private boolean hasNonAscii;
+
+    @Override
+    public void close() {
+    }
+
+    public long getBufferAddress() {
+        return bufAt;
+    }
+
+    public ProtoEntity getEntity(int n) {
+        assert n < nEntities;
+        return entityCache.get(n);
+    }
+
+    public ErrorCode getErrorCode() {
+        return errorCode;
+    }
+
+    public DirectByteCharSequence getMeasurementName() {
+        return measurementName;
+    }
+
+    public long getTimestamp() {
+        return timestamp;
+    }
+
+    public int getnEntities() {
+        return nEntities;
+    }
+
+    public boolean hasNonAsciiChars() {
+        return hasNonAscii;
+    }
+
+    public boolean hasTimestamp() {
+        return timestamp != NULL_TIMESTAMP;
+    }
+
+    public LineTcpParser of(long bufLo) {
+        this.bufAt = bufLo - 1;
+        startNextMeasurement();
+        return this;
+    }
+
+    public ParseResult parseMeasurement(long bufHi) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 23. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.